### PR TITLE
Adding missing types

### PR DESCRIPTION
**Pull request by elopezanaya**
# PR Details

### Description

Missing types definition in public exposed API methods for the SDK, using `any` by default, even for those that have an explicit type to be returned.

_Include a description of the problem to be solved_
## Solution Proposed
_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_

- Adding types as response object for API methods that return a type defined, but for some reason the response is set as `any`
- Defining new types for those methods that doesn't have a type, or are using `any` because the type to return varies on the flow.
- New types are oriented to be frictionless to avoid breaking changes

### Acceptance criteria
_Define what are the conditions to consider the PR has achieved the intended goal_

- API methods exposed should no return `any`
## Test cases and evidence
_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

- SDK validated in OCW
- SDK Validated in OOB
- unit test passing 
- Integration tests passing
image
image

### Sanity Tests

- [ ] You have tested all changes in Popout mode
- [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
- [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y

- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

***Please provide justification if any of the validations has been skipped.***

### CODE REVIEW

File: src/OmnichannelChatSDK.ts [hunk 1]
```diff
@@ -1130,30 +1134,37 @@ class OmnichannelChatSDK {
         }
     }
 
-    public async getDataMaskingRules(): Promise<any> {  // eslint-disable-line  @typescript-eslint/no-explicit-any
+    public async getDataMaskingRules(): Promise<MaskingRules> {  // eslint-disable-line  @typescript-eslint/no-explicit-any
```
Comment (dleblondMSFT) [inline src/OmnichannelChatSDK.ts:1137]: delete the eslint override

File: src/OmnichannelChatSDK.ts [hunk 2]
```diff
@@ -2158,6 +2173,67 @@ class OmnichannelChatSDK {
         }
     }
 
+    private async setPrechatConfigurations(liveWSAndLiveChatEngJoin: LiveWSAndLiveChatEngJoin) {
```
Comment (dleblondMSFT) [inline src/OmnichannelChatSDK.ts:2176]: most of these functions seem to be Promise<void> return type, can we declare it explicitly instead of relying on implicit return type? This prevents developers from accidentally changing it.

File: src/OmnichannelChatSDK.ts [hunk 3]
```diff
@@ -2182,6 +2263,11 @@ class OmnichannelChatSDK {
                 throw error // Bubble up error by default to throw ChatConfigRetrievalFailure
             }
         }
+        return this.liveChatConfig;
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private async buildConfigurations(liveChatConfig: any) {
```
Comment (dleblondMSFT) [inline src/OmnichannelChatSDK.ts:2270]: Can this one be typed?

File: src/types/config.d.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    msdyn_maskforagent: boolean;
+    "@odata.etag": string;
+}
+
+export type DataMaskingInfo = {
+    dataMaskingRules: DataMaskingRules;
+    setting: Setting;
+}
+
+export type LiveWSAndLiveChatEngJoin = {
```
Comment (dleblondMSFT) [inline src/types/config.d.ts:36]: In this type I see many references to | null, is there a reason you've selected this instead of | undefined?

File: src/OmnichannelChatSDK.ts [hunk 5]
```diff
@@ -1780,8 +1794,9 @@ class OmnichannelChatSDK {
 
             const fileManager = new AMSFileManager(this.AMSClient as FramedClient, this.acsAdapterLogger, options);
             return createACSAdapter(optionalParams, this.chatSDKConfig, this.liveChatVersion, ChatAdapterProtocols.ACS, this.telemetry as typeof AriaTelemetry, this.scenarioMarker, this.omnichannelConfig, this.chatToken, fileManager, this.ACSClient?.getChatClient() as ChatClient, this.acsAdapterLogger as ACSAdapterLogger);
+
         } else if (protocol === ChatAdapterProtocols.IC3 || this.liveChatVersion === LiveChatVersion.V1) {
-            return createIC3Adapter(optionalParams, this.chatSDKConfig, this.liveChatVersion, ChatAdapterProtocols.IC3, this.telemetry as typeof AriaTelemetry, this.scenarioMarker, this.chatToken, this.IC3Client, this.ic3ClientLogger as IC3ClientLogger);
+            return  createIC3Adapter(optionalParams, this.chatSDKConfig, this.liveChatVersion, ChatAdapterProtocols.IC3, this.telemetry as typeof AriaTelemetry, this.scenarioMarker, this.chatToken, this.IC3Client, this.ic3ClientLogger as IC3ClientLogger);
```
Comment (matan-hershberg-nuance) [inline src/OmnichannelChatSDK.ts:1799]:
```suggestion
            return createIC3Adapter(optionalParams, this.chatSDKConfig, this.liveChatVersion, ChatAdapterProtocols.IC3, this.telemetry as typeof AriaTelemetry, this.scenarioMarker, this.chatToken, this.IC3Client, this.ic3ClientLogger as IC3ClientLogger);
```

File: src/OmnichannelChatSDK.ts [hunk 6]
```diff
@@ -2191,51 +2277,15 @@ class OmnichannelChatSDK {
...
-        this.liveChatConfig = liveChatConfig;
-        return this.liveChatConfig;
+        Promise.all([
+            this.setDataMaskingConfiguration(dataMaskingConfig),
+            this.setPrechatConfigurations(liveWSAndLiveChatEngJoin),
+            this.setAuthSettingConfig(authSettings),
+            this.setPersistentChatConfiguration(liveWSAndLiveChatEngJoin),
+            this.setCallingOptionConfiguration(liveWSAndLiveChatEngJoin),
+            this.setLocaleIdConfiguration(chatWidgetLanguage),
+            this.setLiveChatVersionConfiguration(liveChatVersion)
```
Comment (matan-hershberg-nuance) [inline src/OmnichannelChatSDK.ts:2287]: 👍

File: src/types/config.d.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    ChatWidgetLanguage: ChatWidgetLanguage;
+}
+
+export type ChatWidgetLanguage = {
+    msdyn_localeid: string,
+    msdyn_languagename: string;
+}
+
+export type DataMaskingRules = {
+    [key: string]: string;
```
Comment (matan-hershberg-nuance) [inline src/types/config.d.ts:22]: Does this need to be `rules: MaskingRule[]` like in the other spots?

File: src/types/response.d.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	rules: MaskingRule[];
+}
+export type GeneralResponse = {} | string | undefined;
+export type GetPreChatSurveyResponse = GeneralResponse;
+export type GetLiveChatTranscriptResponse = GeneralResponse;
+export type DirectLineAdapter = ReturnType<typeof createDirectLine>;
+export type ACSAdapter = ReturnType<typeof createACSAdapter>;
+export type IC3Adapter = ReturnType<typeof createIC3Adapter>;
+export type ChatAdapter = DirectLineAdapter | ACSAdapter | IC3Adapter;
+export type GetAgentAvailabilityResponse = QueueAvailability | undefined;
```
[the remark this answers is not in the document]
Comment (elopezanaya) [reply src/types/response.d.ts:23]: its a convention, because only refers to the action performed

File: src/types/config.d.ts (hunk 4 above)
Comment (elopezanaya) [reply src/types/config.d.ts:40]: I have to confess this was defined by copilot based on the json response from getchatconfig

File: src/OmnichannelChatSDK.ts (hunk 3 above)
Comment (elopezanaya) [reply src/OmnichannelChatSDK.ts]: this one is used only internally, and it takes the object as is defined from OCSDK, which defines the return as   
`public async getChatConfig(requestId: string, bypassCache = false): Promise<object | void>`

**Review by matan-hershberg-nuance — APPROVED** [review]

File: src/types/config.d.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    LiveChatVersion: number;
+    ChatWidgetLanguage: ChatWidgetLanguage;
+}
+
+export type ChatWidgetLanguage = {
+    msdyn_localeid: string,
+    msdyn_languagename: string;
+}
+
+export type DataMaskingRules = {
```
Comment (mithunjohny) [inline src/types/config.d.ts:25]: Can we use DataMaskingRules = Record<string, 'MASKED' | 'UNMASKED'> for objects

**Review by mithunjohny — APPROVED** [review]
